connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/UplynkAdIntegration.kt

```diff
@@ -6,28 +6,57 @@ import com.theoplayer.android.api.ads.ServerSideAdIntegrationHandler
 import com.theoplayer.android.api.player.Player
 import com.theoplayer.android.api.source.SourceDescription
 import com.theoplayer.android.connector.uplynk.UplynkSsaiDescription
-import com.theoplayer.android.connector.uplynk.network.UplynkApi
+import com.theoplayer.android.connector.uplynk.internal.network.UplynkApi
 
 internal class UplynkAdIntegration(
-    val theoplayerView: THEOplayerView,
-    val controller: ServerSideAdIntegrationController,
-    val uplynkDescriptionConverter: UplynkSsaiDescriptionConverter
+    private val theoplayerView: THEOplayerView,
+    private val controller: ServerSideAdIntegrationController,
+    private val eventDispatcher: UplynkEventDispatcher,
+    private val uplynkDescriptionConverter: UplynkSsaiDescriptionConverter,
+    private val uplynkApi: UplynkApi
 ) : ServerSideAdIntegrationHandler {
 
     private val player: Player
         get() = theoplayerView.player
 
-    private val uplynkApi = UplynkApi()
-
     override suspend fun setSource(source: SourceDescription): SourceDescription {
-        val uplynkSource = source.sources.find { it.ssai is UplynkSsaiDescription } ?: return source
-        val ssaiDescription = uplynkSource.ssai as? UplynkSsaiDescription ?: return source
-        val response = uplynkApi.preplay(uplynkDescriptionConverter.buildPreplayUrl(ssaiDescription))
+
+        val uplynkSource = source.sources.singleOrNull { it.ssai is UplynkSsaiDescription }
+        val ssaiDescription = uplynkSource?.ssai as? UplynkSsaiDescription ?: return source
+
+        val response = uplynkDescriptionConverter
+            .buildPreplayUrl(ssaiDescription)
+            .let { uplynkApi.preplay(it) }
+            .also {
+                try {
+                    eventDispatcher.dispatchPreplayEvents(it.parseExternalResponse())
+                } catch (e: Exception) {
+                    eventDispatcher.dispatchPreplayFailure(e)
+                    controller.error(e)
+                }
+            }
```

Comment:
Shouldn't the `uplynkApi.preplay(it)` call be inside the `try..catch`? Otherwise, we're only dealing with *parse* errors, not *network* errors.

Reply:
Yes, that was the idea. It's documented like this:

```

    /**
     * Called when a preplay response is received from Uplynk and failed to be parsed
     *
     * @param exception the `Exception` occurred during the response parsing
     */
    fun onPreplayFailure(exception: Exception) {}
```
Network errors cause a fatal error of the player because any exception in the `setSource` method causes fatal error:

```
     /**
...
     *
     * - If this handler throws an error, the player fires a fatal [ErrorEvent]
     *   (as if by calling [ServerSideAdIntegrationController.fatalError]).
     *
     * @param source
     */
    suspend fun setSource(source: SourceDescription): SourceDescription = source
```

I could change this part in three ways:
1. Keep it as is
2. Call `onPreplayFailure` also for network errors
3. Rename it to `onPreplayDeserialisationFailure`

Reply:
Fair enough.